# I Solving the Sum of Two Digits Programming Challenge

## C++

In [ ]:
%%writefile APlusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Writing APlusB.cpp


In [ ]:
!g++ APlusB.cpp -o APlusB
!./APlusB

4 11
15

### Test input output files

In [ ]:
%%writefile input1.txt
4 8

Writing input1.txt


In [ ]:
%%writefile output1.txt
12

Writing output1.txt


In [ ]:
%%writefile input2.txt
-5 9

Writing input2.txt


In [ ]:
%%writefile output2.txt
4

Writing output2.txt


In [ ]:
!./APlusB < input1.txt

12

In [ ]:
!./APlusB < input2.txt

4

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [ ]:
!grep -xvFf output1.txt output1.txt 

In [ ]:
!grep -xvFf output1.txt output2.txt 

4


In [ ]:
!grep -xvFf output2.txt output1.txt 

12


In [ ]:
!./APlusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
!./APlusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [ ]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Writing sum.py


In [ ]:
!python sum.py

32 90
122


In [ ]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [ ]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

In [ ]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Writing max_pairwise_product.cpp


In [ ]:
!g++ max_pairwise_product.cpp -o max_pairwise_product
!./max_pairwise_product

6
23 1 89 56 34 5
4984


### Test 1: Small numbers

In [ ]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Overwriting input1.txt


In [ ]:
!./max_pairwise_product < input1.txt

90


### Test 2: largest input numbers 200000 200000 in the input

In [ ]:
%%writefile input2.txt
6
1 3 200000 9 200000 23

Overwriting input2.txt


In [ ]:
!./max_pairwise_product < input2.txt

1345294336


Changing the variables to long long

In [ ]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
			       aux_product =  (long long) numbers[first] * numbers[second];
            if ( max_product < aux_product)  {
				        max_product = aux_product;
             }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Writing max_pairwise_product_long.cpp


In [ ]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!./max_pairwise_product_long < input1.txt
!./max_pairwise_product_long < input2.txt


90
40000000000


### Test 3  stress tests 20  times  random n in [2, 20] 

Write the input:


In [ ]:
%%writefile StressTestWr.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
 
  ofstream MyFile("inputStress.txt");


 int number;
 int random;

 for (int i=1; i < 20+1; i++){
     number=2+ rand()% (18+1);
     MyFile << number<<"\n";
     for(int j=1;j<number+1;j++){
      random=1 + rand() % (100000 - 1);
      MyFile << random <<" ";
      }
      MyFile<<"\n";
 }
  
 
  MyFile.close();
}

Overwriting StressTestWr.cpp


In [ ]:
!g++ StressTestWr.cpp -o StressTestWr
!./StressTestWr

Using the naive solution with long long


In [ ]:
%%writefile max_pairwise_product_long_StressTest.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
			       aux_product =  (long long) numbers[first] * numbers[second];
            if ( max_product < aux_product)  {
				        max_product = aux_product;
             }
        }
    }

    return max_product;
}

int main() {
    int n;
    for(int s=1;s<20+1;s++){
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";}
    return 0;
}


Overwriting max_pairwise_product_long_StressTest.cpp


In [ ]:
!g++ max_pairwise_product_long_StressTest.cpp -o max_pairwise_product_long_StressTest
!./max_pairwise_product_long_StressTest < inputStress.txt

3642211002
7128119150
9113410790
9343859618
8884259549
1838756736
9941374800
6975914880
1833433086
4215247776
6828158290
6629084944
9104014296
4740015723
7739814292
9629203782
7398480256
5812280121
6446352898
6458462445


### Test 4  time test  measure the execution time running with the the largest worst case file (200000 increasing numbers, 200000 and 200000 the end) 

In [ ]:
%%writefile Test2Write.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input3.txt");

 MyFile << 200000 <<"\n";  // n 
 for (int i=1; i < 200001; i++){
     MyFile << i <<" ";
 }
  
  // Close the file
  MyFile.close();
}

Writing Test2Write.cpp


In [ ]:
!g++ Test2Write.cpp -o Test2Write
!./Test2Write

Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [ ]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!time ./max_pairwise_product_long < input3.txt

39999800000

real	2m7.215s
user	2m6.997s
sys	0m0.025s


##### Calculate  the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line 

```cpp
 int main() {                         //N inst exec by line executed  
   
    int n;                            //1                      
    std::cin >> n;                    //1
                                              
    std::vector<int> numbers(n);      //n    init vector
    .
    .
 }

```

In [ ]:
%%writefile test_init_vectors.cpp
#include <iostream>
#include <vector>
#include <algorithm>
int main() {                                                               
    int n = 10;                                                                                                                            //1
    std::vector<int> numbers(n);                                            
    for (int i = 0; i < n; ++i) {                                           
        std::cout << numbers[i];                                             
    }

    std::cout << "\n";                       
    return 0;                                                              
}  

Writing test_init_vectors.cpp


In [ ]:
!g++ test_init_vectors.cpp -o test_init_vectors
!./test_init_vectors

0000000000


In [ ]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //N inst exec by line executed 
    long long max_product = 0LL;                                             //1                                          
	  long long aux_product = 0LL;                                             //1 
    int n = numbers.size();                                                  //1 
                                                                             
    for (int first = 0; first < n; ++first) {                                //1+2n  
        for (int second = first + 1; second < n; ++second) {                 //n + (n-1)n 
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2 
            if (max_product < aux_product)  {                                //(n-1)n/2 
				        max_product = aux_product;                                   //[0, (n-1)n/2] 
             }
        }
    }
                                                                             // -------------- 
    return max_product;                                                      // 4 + 3n + [2(n-1)n, 2.5 (n-1)n] 
}

int main() {                                              //N inst exec by line executed 
    int n;                                                //1  
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //1 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 4 + 3n + [2(n-1)n, 2.5 (n-1)n]
    return 0;                                             // -------------- 
}                                                         // 9 + 7n + [2(n-1)n, 2.5 (n-1)n]


Overwriting max_pairwise_product_long.cpp


Count executed intructions in code and verify with inputs 1, 2 and 3

In [ ]:
%%writefile max_pairwise_product_long_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {              //N inst exec by line executed 
    
    long long m = 0LL;                                                       //1
    
    long long max_product = 0LL;                                             //1
    execints++;

	  long long aux_product = 0LL;                                             //1
    execints++;

    int n = numbers.size();                                                  //1 
    execints++;

    execints++;                                                                         
    for (int first = 0; first < n; ++first) {                                //1+2n  
        execints += 2;

        long long m = 0LL;

        execints++;                                    
        for (int second = first + 1; second < n; ++second) {                 //n+ (n-1)n 
            execints += 2;                                                

			      aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2 
            execints++; 

            execints++; 
            if ( max_product < aux_product)  {                               //(n-1)n/2 
                       
				        max_product = aux_product;                                   //[0, (n-1)n/2] 
                execints++; 
                m++;
             }
        }
        if ( (first <= 10) || (n- first <= 10)) std::cout << m << "\n";
    }
                                                                            // -------------- 
    return max_product;                                                     // 4 + 3n + [2(n-1)n, 2.5 (n-1)n] 
}

int main() {                                                 //N inst exec by line executed 
    long long execints =0LL;

    int n;                                                   //1
    execints++;                                                              
    
    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;
    
    execints++;
    for (int i = 0; i < n; ++i) {                            //1 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n
    }
    
    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 4 + 3n + [2(n-1)n, 2.5 (n-1)n]

    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout << "Total:"<< execints << "\n";                       
    return 0;                                                // -------------- 
}                                                            // 9 + 7n + [2(n-1)n, 2.5 (n-1)n]

Overwriting max_pairwise_product_long_execints.cpp


In [ ]:
!g++ max_pairwise_product_long_execints.cpp -o max_pairwise_product_long_execints
!./max_pairwise_product_long_execints < input1.txt
!./max_pairwise_product_long_execints < input2.txt
!./max_pairwise_product_long_execints < input3.txt

9
5
4
3
2
2
2
2
1
0
Total:289
2
1
2
0
0
0
Total:116
199999
100000
66667
50000
40000
33334
28572
25000
22223
20000
18182
2
2
2
2
2
2
2
2
1
0
Total:80003672077


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T = 9 + 7n + [2(n-1)n, 2.5 (n-1)n]$$

It depends on the size of the lists and  the order of the list.

For the input1.txt 
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 9 + 7(10) + 2(9)(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 289 $$

For the input2.txt 
```
6
1 3 200000 9 200000 23
```
$$T = 9 + 7(6) + 2(5)(6)  +  2 + 1 + 2 + 0 + 0 + 0 = 116 $$


For the input3.txt that is the worst case (n=200000 and list in incresing order)

```
200000
1 2 3 4 5 6 7 8 9 10 ... 199998 199999 200000
```

$$
\begin{array}{ll}
T &= 9 + 7(200000) + (2)(199999)(200000)  + (199999 +
100000 + 66667 +50000 \cdots 2 + 2 + 1)\\
&= 80003672077  \\
  &< 9 + 7(200000) + (2.5)(199999)(200000) \\
  & = 100000900009
\end{array}
  $$

Calculate estimated time by instruction and number of instructions by second

In [ ]:
# Estimated time by instruction

128/80003672077 


1.5999265618308827e-09

In [ ]:
# Estimated number of instructions by second

1/(128/80003672077) 

625028688.1015625

### Efficient solution

Add in each line a coment with the number of intstrcution that are executed in that line

In [ ]:
%%writefile efficient.cpp 
#include <iostream>
#include <vector>
#include <algorithm>


long long MaxPairwiseProductFast(const std::vector<int>& numbers) {
    long long max_product = 0;                                             //1
    int n = numbers.size();                                                //1
    
    int max_index1=-1;                                                    //1
    int max_index2=-1;                                                    //1

    for (int first = 0; first < n; ++first) {                            //1+2n  
        if((max_index1==-1)||(numbers[first]>numbers[max_index1])){       // n
        	max_index1=first;                                               //[0,n]
		}
    }
    
    
    for (int second=0; second<n;++second){                                                   //1+2n
    	if((second!=max_index1)&&((max_index2==-1)||(numbers[second]>numbers[max_index2]))){   // n
        	max_index2=second;                                                                //[0,n]
		}
	}
    
    max_product=((long long)(numbers[max_index1]))* ((long long) numbers[max_index2]);      //1

    return max_product;                                                                     //7+6n + [0,2n]
}



int main() {
	
    int n;                                                   //1
    std::cin >> n;                                           //1
    std::vector<int> numbers(n);                            //n
    for(int i=0;i<n; ++i){                                  //1+2n
    	std::cin>> numbers[i];                                //n
	}
	long long result=MaxPairwiseProductFast(numbers);       //1+1+7+6n + [0,2n]                                 
	std::cout<<result<<"\n";
    return 0;                                             // 12+10n+[0,2n]
}


Writing efficient.cpp


Test with inputs 1 and 2



In [ ]:
!g++ efficient.cpp -o efficient
!./efficient < input1.txt

90


In [ ]:
!g++ efficient.cpp -o efficient
!./efficient < input2.txt

40000000000


Test and time it with input 3

In [ ]:
!time ./efficient< input3.txt

39999800000

real	0m0.063s
user	0m0.057s
sys	0m0.004s


20 stress tests

In [ ]:
%%writefile efficientStress.cpp 
#include <iostream>
#include <vector>
#include <algorithm>


long long MaxPairwiseProductFast(const std::vector<int>& numbers) {
    long long max_product = 0;
    int n = numbers.size();
    
    int max_index1=-1;
    int max_index2=-1;

    for (int first = 0; first < n; ++first) {
        if((max_index1==-1)||(numbers[first]>numbers[max_index1])){
        	max_index1=first;
		}
    }
    
    
    for(int second=0; second<n;++second){
    	if((second!=max_index1)&&((max_index2==-1)||(numbers[second]>numbers[max_index2]))){
        	max_index2=second;
		}
	}
    
    max_product=((long long)(numbers[max_index1]))* ((long long) numbers[max_index2]);

    return max_product;
}



int main() {

	
    int n;
  for(int s=1;s<21;s++){
    std::cin >> n;
    std::vector<int> numbers(n);
    for(int i=0;i<n; ++i){
    	std::cin>> numbers[i];
	  }
    long long result=MaxPairwiseProductFast(numbers);
    std::cout<<result<<"\n";
  }

    return 0;
}


Writing efficientStress.cpp


In [ ]:
!g++ efficientStress.cpp -o efficientStress
!./efficientStress < inputStress.txt

3642211002
7128119150
9113410790
9343859618
8884259549
1838756736
9941374800
6975914880
1833433086
4215247776
6828158290
6629084944
9104014296
4740015723
7739814292
9629203782
7398480256
5812280121
6446352898
6458462445


In [ ]:
!./max_pairwise_product_long_StressTest < inputStress.txt > outputStr1.txt
!./efficientStress < inputStress.txt > outputStr2.txt

Compare Stress Test with grep: Long long naive solution and Efficient Solution

In [ ]:
!grep -xvFf outputStr1.txt outputStr2.txt

Count executed instructions in code and verify with inputs 1,2 and 3


In [ ]:
%%writefile efficientExecints.cpp 
#include <iostream>
#include <vector>
#include <algorithm>


long long MaxPairwiseProductFast(const std::vector<int>& numbers, long long &execints ) {
    long long max_product = 0;                                             //1
    execints++;

    int n = numbers.size();                                                //1
    execints++;

    int max_index1=-1;                                                    //1
    execints++;

    int max_index2=-1;                                                    //1
    execints++;


    execints++;
    for (int first = 0; first < n; ++first) {                            //1+2n  
        execints+=2;

        execints++;
        if((max_index1==-1)||(numbers[first]>numbers[max_index1])){       // n
                                                              

        	max_index1=first;                                               //[0,n]
          execints++;


		}
    }
    
    execints++;
    for (int second=0; second<n;++second){                                                   //1+2n
      execints+=2;

      execints++;
    	if((second!=max_index1)&&((max_index2==-1)||(numbers[second]>numbers[max_index2]))){   // n
        	max_index2=second;                                                                //[0,n]
          execints++;
		}
	}


    execints++;
    max_product=((long long)(numbers[max_index1]))* ((long long) numbers[max_index2]);      //1

    return max_product;                                                                     //7+6n + [0,2n]
}



int main() {
    long long execints=0LL;
	
    int n;                                                   //1
    execints++;

    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                            //n
    execints+=n;


    execints++;
    for(int i=0;i<n; ++i){                                  //1+2n
      execints+=2;            
    	std::cin>> numbers[i];                                //n
      execints++;
	}
	long long result=MaxPairwiseProductFast(numbers,execints);       //1+1+7+6n + [0,2n]  
  execints++;                               
	//std::cout<<result<<"\n";
  
  execints++;

  std::cout<<"Total: "<<execints<<"\n";
    return 0;                                             // 12+10n+[0,2n]
}


Writing efficientExecints.cpp


In [ ]:
!g++ efficientExecints.cpp -o efficientExecints
!./efficientExecints < input1.txt
!./efficientExecints < input2.txt
!./efficientExecints < input3.txt

Total: 131
Total: 79
Total: 2400011


Calculate estimated time by instruction and number of instructions by second

In [ ]:
# Estimated time by instruction

0.057/2400011

2.3749891146332247e-08

In [ ]:
# Estimated number of instructions by second

1/(0.057/2400011)

42105456.14035088

### Super efficient solution

Whitout storing the list in memory and whitout calling a function, add in each line a coment with the number of intstrcution that are executed in that line 

In [ ]:
%%writefile superEfficient.cpp
#include <iostream>
#include <vector>
#include <algorithm>


int main() {

	
    int n;                                                             //1
    std::cin >> n;                                                     //1
    long long max_numb = 0LL;                                          //1
    long long sec_max_numb = 0LL;                                      //1
    long long temporal=0LL;                                            //1
    long long max_product=0LL;                                          //1


    for(int i=0;i<n; ++i){                                            //1+2n
        
    	std::cin>> temporal;                                             //n

      if(temporal>=max_numb){                                          //n
          sec_max_numb=max_numb;                                      //[0,n]
          max_numb=temporal;                                          //[0,n]
      }
      else if(temporal>sec_max_numb){                                 //[0,n]
          sec_max_numb=temporal;                                      //[0,n]
      }
	  }


    max_product=((long long)max_numb) * ((long long) sec_max_numb);   //1

  	std::cout<<max_product<<"\n";                                      //1
    return 0;                                                         //9+4n+[0,2n]
    
}

Writing superEfficient.cpp


Test with inputs 1 and 2


 

In [ ]:
!g++ superEfficient.cpp -o superEfficient
!./superEfficient < input1.txt
!./superEfficient < input2.txt

90
40000000000


Test and time it with input 3

In [ ]:
!time ./superEfficient < input3.txt

39999800000

real	0m0.060s
user	0m0.055s
sys	0m0.003s


20 stress tests 

In [ ]:
%%writefile superEfficientStress.cpp
#include <iostream>
#include <vector>
#include <algorithm>


int main() {

	
    int n;  
    for(int s=1;s<21;s++)                                                           
    {  std::cin >> n;                                                     
      long long max_numb = 0LL;                                          
      long long sec_max_numb = 0LL;                                      
      long long temporal=0LL;                                            
      long long max_product=0LL;                                          


      for(int i=0;i<n; ++i){                                            
          
        std::cin>> temporal;                                             

        if(temporal>=max_numb){                                          
            sec_max_numb=max_numb;                                      
            max_numb=temporal;                                          
        }
        else if(temporal>sec_max_numb){                                 
            sec_max_numb=temporal;                                      
        }
      }


      max_product=((long long)max_numb) * ((long long) sec_max_numb);   

      std::cout<<max_product<<"\n";       }                              
    return 0;                                                         
    
}

Writing superEfficientStress.cpp


In [ ]:
!g++ superEfficientStress.cpp -o superEfficientStress
!./superEfficientStress < inputStress.txt

3642211002
7128119150
9113410790
9343859618
8884259549
1838756736
9941374800
6975914880
1833433086
4215247776
6828158290
6629084944
9104014296
4740015723
7739814292
9629203782
7398480256
5812280121
6446352898
6458462445


In [ ]:
!./superEfficientStress < inputStress.txt >outputStr3.txt

In [ ]:
!grep -xvFf outputStr1.txt outputStr3.txt

Count executed instructions in code and verify with inputs 1,2 and 3

In [ ]:
%%writefile superEfficientExecints.cpp
#include <iostream>
#include <vector>
#include <algorithm>


int main() {

    long long execints=0LL;
    int n;                                                             //1
    execints++;
    std::cin >> n;                                                     //1
    execints++;
    long long max_numb = 0LL;                                          //1
    execints++;
    long long sec_max_numb = 0LL;                                      //1
    execints++;
    long long temporal=0LL;                                            //1
    execints++;
    long long max_product=0LL;                                          //1
    execints++;

    execints++;
    for(int i=0;i<n; ++i){                                            //1+2n
    execints+=2;
        
    	std::cin>> temporal;                                             //n
      execints++;

      execints++;
      if(temporal>=max_numb){                                          //n
          sec_max_numb=max_numb;                                      //[0,n]
          execints++;
          max_numb=temporal;                                          //[0,n]
          execints++;
      }
      else if(temporal>sec_max_numb){                                 //[0,n]
            execints++;
          sec_max_numb=temporal;                                      //[0,n]
          execints++;
      }
	  }


    max_product=((long long)max_numb) * ((long long) sec_max_numb);   //1
    execints++;

  	//std::cout<<max_product<<"\n";                                      //1
    execints++;
    std::cout<<"Total: "<<execints<<"\n";   
    return 0;                                                         //9+4n+[0,2n]
    
}

Writing superEfficientExecints.cpp


In [ ]:
!g++ superEfficientExecints.cpp -o superEfficientExecints
!./superEfficientExecints < input1.txt
!./superEfficientExecints < input2.txt
!./superEfficientExecints < input3.txt

Total: 69
Total: 43
Total: 1200009


Calculate estimated time by instruction and number of instructions by second

In [ ]:
# Estimated time by instruction

0.055/1200009


4.583298958591144e-08

In [ ]:
# Estimated number of instructions by second

1/(0.055/1200009)

21818345.454545453

## Python

Efficient cython file reading, string parsing, and array building

https://stackoverflow.com/questions/37219875/efficient-cython-file-reading-string-parsing-and-array-building



File PreProcessing With C and Cython

https://github.com/EveryTimeIWill18/Cython_Repo/blob/master/FastFileProcessingWithCython.ipynb



### Efficient solution

In [ ]:
%%writefile max_pairwise_product.py 

if __name__ == '__main__':
    input_n = int(input())
    input_numbers = [int(x) for x in input().split()]
    max_numb=0
    second_max_numb=0
    max_product=0
    for temporal in range(input_n):
      if (input_numbers[temporal]>=max_numb):
        second_max_numb=max_numb
        max_numb=input_numbers[temporal]
      elif (input_numbers[temporal]>second_max_numb):
        second_max_numb=input_numbers[temporal]

    max_product= max_numb * second_max_numb
    print(max_product)

Writing max_pairwise_product.py


In [ ]:
!python max_pairwise_product.py

4
1 3 2 3
9


In [ ]:
!python max_pairwise_product.py < input1.txt

90


In [ ]:
!python max_pairwise_product.py < input2.txt

40000000000


In [ ]:
!time python max_pairwise_product.py < input3.txt

39999800000

real	0m0.175s
user	0m0.134s
sys	0m0.036s


## StressTest

In [ ]:
%%writefile max_pairwise_productStress.py 

if __name__ == '__main__':
  for s in range(1,21):
      input_n = int(input())
      input_numbers = [int(x) for x in input().split()]
      max_numb=0
      second_max_numb=0
      max_product=0
      for temporal in range(input_n):
        if (input_numbers[temporal]>=max_numb):
          second_max_numb=max_numb
          max_numb=input_numbers[temporal]
        elif (input_numbers[temporal]>second_max_numb):
          second_max_numb=input_numbers[temporal]

      max_product= max_numb * second_max_numb
      print(max_product)

Writing max_pairwise_productStress.py


In [ ]:
!python max_pairwise_productStress.py < inputStress.txt

3642211002
7128119150
9113410790
9343859618
8884259549
1838756736
9941374800
6975914880
1833433086
4215247776
6828158290
6629084944
9104014296
4740015723
7739814292
9629203782
7398480256
5812280121
6446352898
6458462445


In [ ]:
!python max_pairwise_productStress.py < inputStress.txt> outputStrPy.txt

In [ ]:
!grep -xvFf outputStr3.txt outputStrPy.txt

# Screenhsoots of the Coursera Submissions in C++ and Python

![AplusB](https://github.com/geramirezl/AlgorithmsUN2021I/blob/main/Lab6/AplusB.jpg?raw=true)

![MaximumPairwiseProduct](https://github.com/geramirezl/AlgorithmsUN2021I/blob/main/Lab6/MaximumPairwiseProduct.jpg?raw=true)